### ISSUE TITLE
Making a 3D game using Pyglet

### ISSUE THREAD

**Issue by eschan145** [pyglet/pyglet#713]
I've been using Pyglet for a couple of years now (along with arcade), and I am looking into creating a 3D game. Is pyglet designed for such a game? My game will have to have 70K+ sprites moving at high speed and very many calculations. I would need **very good** performance. I know there are other libraries out there, but I kind of want to stick with pyglet.

**Comment by benmoran56** [comment]
Hi, thanks for opening the ticket.

This will greatly depend on the type of game, and how you write it. If you need 70K+ sprites, then it's probably not possible to do that purely on the CPU. 

Do you have an example of what type of game you're wanting to make?
What kind of calculations do you mean?

**Comment by eschan145** [comment]
It's a strategical medieval warfare simulator game. Very wordy, but describes it. I would need at least 100K soldiers and sprites. And it can't be laggy.

**Comment by assyrianic** [comment]
honestly this _can_ be done but you'd likely have to do the calculations in C as opposed to Python or use a high-performance, pre-existing library to work it out with Pyglet.

Pyglet can definitely do 3D but would depend on how strong the graphics would be.

**Comment by benmoran56** [comment]
Graphically speaking only: 
You can render that many objects in pyglet. You'll have to do some screen culling, so that you don't waste time drawing off-screen objects, but that's something you'll need to do in any language or framework. 
The weakest link in pyglet comes from updating OpenGL objects from Python. If you need to individually move lots of objects each frame, then you will be limited to several thousand per frame (at 60fps). This promises to be better in Python 3.11, but Python overhead comes into play here. 

I would suggest you create a graphical mock-up, that mimics what you actually want to do in the final game. This will give you an idea of whether you will have enough CPU left over for your logic.

**Comment by eschan145** [comment]
Could you provide an example of doing screen culling of drawing 10K or something objects and still have a high fps?

1. I will need to need to move _thousands_ or objects each frame. From swords to arrows, soldiers, all that stuff.
2. Is it possible to draw only individual objects in a `pyglet.graphics.Batch`? I've tried before, but couldn't find any methods. So basically can an object be removed from a batch?
3. I'll have to do a lot of logic, for like eighty thousand sprites

I'm also familiar with `arcade`. I know it's for creating "2D video games", but this might be an exception. I've never tried loading a 3D model whatsoever, so I don't know the performance issues or stuff.

**Comment by eschan145** [comment]
@assyrianic is there an easy way to import C calculations into Python?

**Comment by shenjackyuanjie** [comment]
> @assyrianic is there an easy way to import C calculations into Python?

you can try [Cython](https://cython.org/) to compile your code into c code and implment it's performence

**Comment by shenjackyuanjie** [comment]
> 3D physics engine

hmmmmmm
as far as I know, you just need a gcc to use Cython, and you can find gcc on mingw64
and you can just use C-defiend library just as native python in Cython

for 3D physics engine, I don't know for sure
so I just go bing it. and found [Panda 3D](https://docs.panda3d.org/1.10/python/programming/physics/builtin/index)

**Comment by iperov** [comment]
use https://www.taichi-lang.org/

**Comment by einarf** [comment]
It's also possible to move some of the costly logic into shaders to get that speed bump, but it can be less flexible and require a fair bit of shader knowledge. In some cases it's trivial.. other.. not so much.

**Comment by shenjackyuanjie** [comment]
> It's also possible to move some of the costly logic into shaders to get that speed bump, but it can be less flexible and require a fair bit of shader knowledge. In some cases it's trivial.. other.. not so much.

yea you can try to move some logic to shader. just like [DSP](https://store.steampowered.com/app/1366540/_/)
they move alot of compute and logic to GPU work, maybe you can try those idea to improve your logic

**Comment by arthurmoreno** [comment]
> Graphically speaking only:
You can render that many objects in pyglet. You'll have to do some screen culling, so that you don't waste time drawing off-screen objects, but that's something you'll need to do in any language or framework.
The weakest link in pyglet comes from updating OpenGL objects from Python. If you need to individually move lots of objects each frame, then you will be limited to several thousand per frame (at 60fps). This promises to be better in Python 3.11, but Python overhead comes into play here.
> I would suggest you create a graphical mock-up, that mimics what you actually want to do in the final game. This will give you an idea of whether you will have enough CPU left over for your logic.

I am doing exactly that right now, I am using numba to do the heavy world state checks and updates (Which is another topic about optimizing the memory transfer between host to device)

I know that I can still do some optimizations in my own drawing method, but the most costly thing at the end is the following:
```
  Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     ...
     2103    0.001    0.000    0.002    0.000 /home/arthur/anaconda3/envs/life-sim/lib/python3.9/site-packages/pyglet/graphics/vertexdomain.py:303(__getattr__)
      512    0.000    0.000    0.001    0.000 /home/arthur/anaconda3/envs/life-sim/lib/python3.9/site-packages/pyglet/sprite.py:728(visible)
      512    0.001    0.000    0.001    0.000 /home/arthur/anaconda3/envs/life-sim/lib/python3.9/site-packages/pyglet/sprite.py:457(_update_position)
     2103    0.001    0.000    0.001    0.000 /home/arthur/anaconda3/envs/life-sim/lib/python3.9/site-packages/pyglet/graphics/vertexbuffer.py:304(invalidate)
      512    0.000    0.000    0.001    0.000 /home/arthur/anaconda3/envs/life-sim/lib/python3.9/site-packages/pyglet/sprite.py:500(x)
      512    0.000    0.000    0.001    0.000 /home/arthur/anaconda3/envs/life-sim/lib/python3.9/site-packages/pyglet/sprite.py:513(y)
     ...
```

There is a way to optimize those pieces of code with any JIT library like numba or even compile some pieces of pyglet into Cython? Does anyone knows of a fork trying to do such a thing?

**Comment by einarf** [comment]
I think @benmoran56 created a GeoSprite somewhere using geometry shader to emit a sprite from a point reducing the vertex data you need to handle by quite a bit. That's worth looking into.

https://github.com/pyglet/pyglet/blob/9c613569c01e761624e78ffa0373f604fdb43014/experimental/geoshader_sprite.py

**Comment by benmoran56** [comment]
Oops, that one is still using deprecated `Batch.add`. 
This branch here has the latest version. The `sprite.py` file should be a drop-in replacement:
https://github.com/pyglet/pyglet/blob/geosprite/pyglet/sprite.py

To take things further, do you need a full Sprite class at all? If you only need simple colored shapes or textured quads (without all of the additional rotation/opacity/scaling functionality, it should be a little bit quicker to just make raw VertexLists using a small helper function. Something along the lines of this example here:
https://github.com/pyglet/pyglet/blob/master/examples/opengl/opengl_shader.py#L112

**Comment by cspotcode** [comment]
Is geosprite out of sync with master?  I've tried using it as a drop-in replacement, but it's raising an error when attempting to access `texture.uv`.  I'm not familiar with the code and would like to avoid going down a rabbit hole.

**Comment by benmoran56** [comment]
@cspotcode Yes, this had fallen out of sync. I updated the master branch with the latest example, which can be found under `experimental/` in the repo.

**Comment by cspotcode** [comment]
Thanks, would it be possible to move those experimental bits under `pyglet/` so that they're usable from a `pip install -e .`?  Or is there another way you'd recommend using them?

I've been tinkering on `GeoSprite`, benchmarking it against `arcade.Sprite`.  It was initially slower than arcade's `Sprite` in my benchmark.  I found that the `VertexDomain` and `VertexList` abstractions are a bit heavy.  But I've made a more optimized path for sprites to write updated values into the underlying buffers without violating the abstractions too much.  This makes `GeoSprite` 1.2x faster than `arcade.Sprite`.  (perhaps next I can optimize arcade's sprite to catch up!)

I'll share what I have in a PR.
